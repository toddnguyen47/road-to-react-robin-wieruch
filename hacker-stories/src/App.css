.App {
  text-align: center;
}

.App-logo {
  height: 40vmin;
  pointer-events: none;
}

@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-spin infinite 20s linear;
  }
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #61dafb;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}

/* o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o- */
/*   new things! */
/* o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o- */
html {
  font-size: 62.5%; /* 10 / 16 = 62.5% */
}

body {
  font-size: 1.6rem;
}

.container {
  /* height: 100vw; */
  padding: 2rem;

  background: #83a4d4; /* fallback for old browsers */
  background: linear-gradient(to left, #b6fbff, #83a4d4);

  color: #171212;
}

.headline-primary {
  font-size: 4.8rem;
  font-weight: 300;
  letter-spacing: 0.2rem;
}

.item {
  display: flex;
  align-items: center;
  padding-bottom: 0.5rem;
}

.item > span {
  padding: 0rem 0.5rem;
  white-space: nowrap;
  overflow: hidden;
  /* white-space: nowrap; */
  text-overflow: ellipsis;
}

.item > span > a {
  color: inherit;
}
