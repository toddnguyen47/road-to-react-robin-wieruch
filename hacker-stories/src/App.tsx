import React from "react";
import axios from "axios";
import classnames from "classnames";

import styles from "./App.module.css";

// o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o
//   CONSTANTS
// o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o
const title = "React";
const welcome = {
  greeting: "Hey",
  title: title,
};

// o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o
//   TYPE & INTERFACE DECLARATIONS
// o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o
type Story = {
  objectID: string;
  url: string;
  title: string;
  author: string;
  num_comments: number;
  points: number;
};

type Stories = Array<Story>;

type OnRemoveItem = (item: Story) => void;

type InputWithLabelProps = {
  id: string;
  type?: string;
  value: string;
  onInputChanged: (event: React.ChangeEvent<HTMLInputElement>) => void;
  outputString: string;
  isFocused?: boolean;
  children: React.ReactNode;
};

type ItemProps = {
  item: Story;
  onRemoveItem: OnRemoveItem;
};

type ListProps = {
  list: Stories;
  onRemoveItem: OnRemoveItem;
};

type SearchFormProps = {
  searchTerm: string;
  onSearchInput: (event: React.ChangeEvent<HTMLInputElement>) => void;
  onSearchSubmit: (event: React.FormEvent<HTMLFormElement>) => void;
};

interface StoriesFetchInitAction {
  type: "STORIES_FETCH_INIT";
}

interface StoriesFetchSuccessAction {
  type: "STORIES_FETCH_SUCCESS";
  payload: Stories;
}

interface StoriesFetchFailureAction {
  type: "STORIES_FETCH_FAILURE";
}

interface StoriesRemoveAction {
  type: "STORIES_REMOVE";
  payload: Story;
}

type StoriesState = {
  data: Stories;
  isLoading: boolean;
  isError: boolean;
};

type StoriesAction =
  | StoriesFetchInitAction
  | StoriesFetchSuccessAction
  | StoriesFetchFailureAction
  | StoriesRemoveAction;

// o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o
//   FUNCTIONS
// o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o

function getTitle(title: string) {
  return "| " + title + " |";
}

// const List = (props) => {
function List({ list, onRemoveItem }: ListProps) {
  return (
    <>
      <div className={styles["item"]}>
        <span style={{ width: "40%" }}>URL</span>
        <span style={{ width: "30%" }}>Author</span>
        <span style={{ width: "10%" }}># Comments</span>
        <span style={{ width: "10%" }}>Points</span>
        <span style={{ width: "10%" }}></span>
      </div>

      <div>
        {list.map((item) => {
          return (
            <Item key={item.objectID} item={item} onRemoveItem={onRemoveItem} />
          );
        })}
      </div>
    </>
  );
}

const Item = ({ item, onRemoveItem }: ItemProps) => {
  return (
    <div className={styles["item"]}>
      <span style={{ width: "40%" }}>
        <a href={item.url}>{item.title}</a>
      </span>
      <span style={{ width: "30%" }}>{item.author}</span>
      <span style={{ width: "10%" }}>{item.num_comments}</span>
      <span style={{ width: "10%" }}>{item.points}</span>
      <span style={{ width: "10%" }}>
        <button
          type="button"
          className={`${styles["button"]} ${styles["button__small"]}`}
          onClick={() => onRemoveItem(item)}
        >
          Dismiss
        </button>
      </span>
    </div>
  );
};

// /** Passing in a full props javascript object */
// const Search = (props) => {

/** Destructuring in the function parameter! */
const InputWithLabel = ({
  id,
  type = "text",
  value,
  onInputChanged,
  outputString,
  isFocused,
  children,
}: InputWithLabelProps) => {
  // (2A) create an inputRef
  const inputRef = React.useRef<HTMLInputElement>(null!);

  // (2C) Go into React's lifecycle. Perform the focus when the element renders or
  // when its dependencies change
  React.useEffect(() => {
    // Check if the `.current` property exists
    if (isFocused && inputRef.current) {
      // (2D) Since `inputRef` is passed into the `input` element as the `ref` attribute,
      // `.current` property gives us access to that `input` element
      inputRef.current.focus();
      console.log("Focusing!");
    }
  }, [isFocused]);

  // (1B) Use the callback function
  return (
    <>
      <label htmlFor={id} className={styles["label"]}>
        {children}
      </label>
      <input
        /** (2B) The `inputRef` variable is passed to JSX's reserved `ref` attribute.
         * The element instance is assigned to the mutable `current` property
         */
        ref={inputRef}
        id={id}
        type={type}
        onChange={onInputChanged}
        value={value}
        className={styles["input"]}
      />

      <p>
        {outputString} <strong>'{value}'</strong>
      </p>
    </>
  );
};

/**
 * Store a value in localStorage.
 */
const useSemiPersistentState = (
  key: string,
  initialState: string
): [string, (newValue: string) => void] => {
  const [value, setValue] = React.useState(
    localStorage.getItem(key) || initialState
  );

  /** We will need to monitor both `key` and `value`
   * In case `key` changes outside this hook
   */
  React.useEffect(() => {
    // Store in localStorage every time searchTerm is changed
    localStorage.setItem(key, value);
  }, [key, value]);

  return [value, setValue];
};

const storiesReducer = (state: StoriesState, action: StoriesAction) => {
  switch (action.type) {
    case "STORIES_REMOVE":
      return {
        ...state,
        data: state.data.filter(
          (story) => action.payload.objectID !== story.objectID
        ),
      };
    case "STORIES_FETCH_INIT":
      return {
        ...state,
        isLoading: true,
        isError: false,
      };
    case "STORIES_FETCH_SUCCESS":
      return {
        ...state,
        isLoading: false,
        isError: false,
        data: action.payload,
      };
    case "STORIES_FETCH_FAILURE":
      return {
        ...state,
        isLoading: false,
        isError: true,
      };
    default:
      throw new Error();
  }
};

const getSumComments = (storiesState: StoriesState): number => {
  const initCommentValue = 0;
  return storiesState.data.reduce(
    (result, value) => result + value.num_comments,
    initCommentValue
  );
};

const SearchForm = ({
  searchTerm,
  onSearchInput,
  onSearchSubmit,
}: SearchFormProps) => {
  return (
    <>
      <form onSubmit={onSearchSubmit} className={styles["searchForm"]}>
        <InputWithLabel
          id="search"
          value={searchTerm}
          onInputChanged={onSearchInput}
          isFocused={true}
          outputString="The Search Term is: "
        >
          <strong>Search: &nbsp;</strong>
        </InputWithLabel>

        <button
          type="submit"
          disabled={!searchTerm}
          className={classnames(styles["button"], styles["button__large"])}
        >
          Submit
        </button>
      </form>
    </>
  );
};

/**
 * This is the `App` component.
 * Everything outside is the global space!
 */
const App = () => {
  const [searchTerm, setSearchTerm] = useSemiPersistentState("searchKey", "");

  // const [stories, setStories] = React.useState([]);
  const [stories, dispatchStories] = React.useReducer(storiesReducer, {
    data: [] as Story[],
    isLoading: false,
    isError: false,
  });

  const sumOfStoriesComments = getSumComments(stories);

  // (3A) fetch popular tech stories
  const API_ENDPOINT = "https://hn.algolia.com/api/v1/search?query=";

  const handleSearchInput = (event: React.ChangeEvent<HTMLInputElement>) => {
    setSearchTerm(event.target.value);
  };

  /** Make a new `url` state and a function to update it */
  const [url, setUrl] = React.useState(`${API_ENDPOINT}${searchTerm}`);

  const handleSearchSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    setUrl(`${API_ENDPOINT}${searchTerm}`);
    event.preventDefault();
  };

  const handleRemoveStory = (item: Story) => {
    dispatchStories({
      type: "STORIES_REMOVE",
      payload: item,
    });
  };

  const handleFetchStories = React.useCallback(async () => {
    // This check isn't needed technically as the button is disabled,
    // but it does not hurt to check
    if (!url) return;

    dispatchStories({ type: "STORIES_FETCH_INIT" });
    try {
      const result = await axios.get(url);
      console.log("Fetching!");
      dispatchStories({
        type: "STORIES_FETCH_SUCCESS",
        payload: result.data.hits,
      });
    } catch {
      dispatchStories({ type: "STORIES_FETCH_FAILURE" });
    }
  }, [url]); // (4E) when `searchTerm` changes

  React.useEffect(() => {
    handleFetchStories(); // (4C) invoke the callback using the useEffect() hook
  }, [handleFetchStories]); // (4D) dependency array: depends on any re-defined function handleFetchStories

  return (
    <div className={styles["container"]}>
      <h1 className={styles["headline-primary"]}>
        {welcome.greeting} {welcome.title}
      </h1>

      <h1 className={styles["headline-primary"]}>Hello {getTitle("React")}</h1>

      <SearchForm
        searchTerm={searchTerm}
        onSearchInput={handleSearchInput}
        onSearchSubmit={handleSearchSubmit}
      />

      <p>Sum of Number of Comments: {sumOfStoriesComments}</p>

      {stories.isError && <p>Cannot retrieve stories data.</p>}

      {stories.isLoading ? (
        <p>Loading...</p>
      ) : (
        <List list={stories.data} onRemoveItem={handleRemoveStory} />
      )}
    </div>
  );
};

export default App;

export { Item, List };
