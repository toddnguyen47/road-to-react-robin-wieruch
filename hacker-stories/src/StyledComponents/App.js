import React from "react";
import axios from "axios";
import styled from "styled-components";

import { ReactComponent as Check } from "../images/check.svg";

// CONSTANTS
const TypeSetEnum = {
  // Do not start enum at 0 because of JavaScript's strange equality comparison
  // when it comes to 0
  REMOVE_STORIES: 1,
  STORIES_FETCH_INIT: 2,
  STORIES_FETCH_SUCCESS: 3,
  STORIES_FETCH_FAILURE: 4,
};

const title = "React";
const welcome = {
  greeting: "Hey",
  title: title,
};

/**
 *
 * @param {String} title
 */
function getTitle(title) {
  return "| " + title + " |";
}

// function List({ list, onRemoveItem }) {
const List = React.memo(({ list, onRemoveItem }) => {
  console.log("B: List");
  console.log(list.length);

  return (
    <>
      <StyledItem>
        <StyledColumn style={{ width: "40%" }}>URL</StyledColumn>
        <StyledColumn style={{ width: "30%" }}>Author</StyledColumn>
        <StyledColumn style={{ width: "10%" }}># Comments</StyledColumn>
        <StyledColumn style={{ width: "10%" }}>Points</StyledColumn>
        <StyledColumn style={{ width: "10%" }}>Dismiss</StyledColumn>
      </StyledItem>

      <div>
        {list.map((item) => {
          return (
            <Item key={item.objectID} item={item} onRemoveItem={onRemoveItem} />
          );
        })}
      </div>
    </>
  );
});

const Item = ({ item, onRemoveItem }) => {
  return (
    <StyledItem>
      <StyledColumn style={{ width: "40%" }}>
        <a href={item.url}>{item.title}</a>
      </StyledColumn>
      <StyledColumn style={{ width: "30%" }}>{item.author}</StyledColumn>
      <StyledColumn style={{ width: "10%" }}>{item.num_comments}</StyledColumn>
      <StyledColumn style={{ width: "10%" }}>{item.points}</StyledColumn>
      <StyledColumn style={{ width: "10%" }}>
        <StyledButtonSmall type="button" onClick={() => onRemoveItem(item)}>
          <Check width="1.8rem" height="1.8rem" />
        </StyledButtonSmall>
      </StyledColumn>
    </StyledItem>
  );
};

// /** Passing in a full props javascript object */
// const Search = (props) => {

/** Destructuring in the function parameter! */
const InputWithLabel = ({
  id,
  type = "text",
  value,
  onInputChanged,
  outputString,
  isFocused,
  children,
}) => {
  // (2A) create an inputRef
  const inputRef = React.useRef();

  // (2C) Go into React's lifecycle. Perform the focus when the element renders or
  // when its dependencies change
  React.useEffect(() => {
    // Check if the `.current` property exists
    if (isFocused && inputRef.current) {
      // (2D) Since `inputRef` is passed into the `input` element as the `ref` attribute,
      // `.current` property gives us access to that `input` element
      inputRef.current.focus();
      console.log("Focusing!");
    }
  }, [isFocused]);

  // (1B) Use the callback function
  return (
    <>
      <StyledLabel htmlFor={id}>{children}</StyledLabel>
      <StyledInput
        /** (2B) The `inputRef` variable is passed to JSX's reserved `ref` attribute.
         * The element instance is assigned to the mutable `current` property
         */
        ref={inputRef}
        id={id}
        type={type}
        onChange={onInputChanged}
        value={value}
      />

      <p>
        {outputString} <strong>'{value}'</strong>
      </p>
    </>
  );
};

/**
 * Store a vallue in localStorage.
 * @param {String} key
 * @param {*} initialState
 * @returns [String, Function]
 */
const useSemiPersistentState = (key, initialState) => {
  const isMounted = React.useRef(false);

  const [value, setValue] = React.useState(
    localStorage.getItem(key) || initialState
  );

  /** We will need to monitor both `key` and `value`
   * In case `key` changes outside this hook
   */
  React.useEffect(() => {
    if (!isMounted.current) {
      isMounted.current = true;
    }
    else {
      console.log("A: useSemiPersistentState() useEffect()");
      // Store in localStorage every time searchTerm is changed
      localStorage.setItem(key, value);
    }
  }, [key, value]);

  return [value, setValue];
};

const storiesReducer = (state, action) => {
  switch (action.type) {
    case TypeSetEnum.REMOVE_STORIES:
      return {
        ...state,
        data: state.data.filter(
          (story) => action.payload.objectID !== story.objectID
        ),
      };
    case TypeSetEnum.STORIES_FETCH_INIT:
      return {
        ...state,
        isLoading: true,
        isError: false,
      };
    case TypeSetEnum.STORIES_FETCH_SUCCESS:
      return {
        ...state,
        isLoading: false,
        isError: false,
        data: action.payload,
      };
    case TypeSetEnum.STORIES_FETCH_FAILURE:
      return {
        ...state,
        isLoading: false,
        isError: true,
      };
    default:
      throw new Error();
  }
};

const SearchForm = ({ searchTerm, onSearchInput, onSearchSubmit }) => {
  return (
    <>
      <StyledSearchForm onSubmit={onSearchSubmit}>
        <InputWithLabel
          id="search"
          value={searchTerm}
          onInputChanged={onSearchInput}
          isFocused={true}
          outputString="The Search Term is: "
        >
          <strong>Search: &nbsp;</strong>
        </InputWithLabel>

        <StyledButtonLarge type="submit" disabled={!searchTerm}>
          Submit
        </StyledButtonLarge>
      </StyledSearchForm>
    </>
  );
};

/**
 * This is the `App` component.
 * Everything outside is the global space!
 */
const App = () => {
  const [searchTerm, setSearchTerm] = useSemiPersistentState("searchKey", "");

  // const [stories, setStories] = React.useState([]);
  const [stories, dispatchStories] = React.useReducer(storiesReducer, {
    data: [],
    isLoading: false,
    isError: false,
  });

  // (3A) fetch popular tech stories
  const API_ENDPOINT = "https://hn.algolia.com/api/v1/search?query=";

  const handleSearchInput = (event) => {
    setSearchTerm(event.target.value);
  };

  /** Make a new `url` state and a function to update it */
  const [url, setUrl] = React.useState(`${API_ENDPOINT}${searchTerm}`);

  const handleSearchSubmit = (event) => {
    setUrl(`${API_ENDPOINT}${searchTerm}`);
    event.preventDefault();
  };

  /** Use a Callback so it only re-renders if one of its dependencies
    * in the dependency array changes
    */
  const handleRemoveStory = React.useCallback((item) => {
    dispatchStories({
      type: TypeSetEnum.REMOVE_STORIES,
      payload: item,
    });
  }, []);

  const handleFetchStories = React.useCallback(async () => {
    // This check isn't needed technically as the button is disabled,
    // but it does not hurt to check
    if (!url) return;

    dispatchStories({ type: TypeSetEnum.STORIES_FETCH_INIT });
    try {
      const result = await axios.get(url);
      console.log("Fetching!");
      dispatchStories({
        type: TypeSetEnum.STORIES_FETCH_SUCCESS,
        payload: result.data.hits,
      });
    } catch {
      dispatchStories({ type: TypeSetEnum.STORIES_FETCH_FAILURE });
    }
  }, [url]); // (4E) when `searchTerm` changes

  React.useEffect(() => {
    handleFetchStories(); // (4C) invoke the callback using the useEffect() hook
  }, [handleFetchStories]); // (4D) dependency array: depends on any re-defined function handleFetchStories

  console.log("B: App");

  return (
    <StyledContainer>
      <HeadlinePrimary>
        {welcome.greeting} {welcome.title}
      </HeadlinePrimary>

      <HeadlinePrimary>Hello {getTitle("React")}</HeadlinePrimary>

      <SearchForm
        searchTerm={searchTerm}
        onSearchInput={handleSearchInput}
        onSearchSubmit={handleSearchSubmit}
      />

      {stories.isError && <p>Cannot retrieve stories data.</p>}

      {stories.isLoading ? (
        <p>Loading...</p>
      ) : (
          <List list={stories.data} onRemoveItem={handleRemoveStory} />
        )}

      <hr />
      <footer>
        Icons made by{" "}
        <a href="https://www.flaticon.com/authors/freepik" title="Freepik">
          Freepik
        </a>{" "}
        from{" "}
        <a href="https://www.flaticon.com/" title="Flaticon">
          {" "}
          www.flaticon.com
        </a>
      </footer>
    </StyledContainer>
  );
};

export default App;

// o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o
//   STYLED COMPONENTS
// o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o-o

const HeadlinePrimary = styled.h1`
  font-size: 4.8rem;
  font-weight: 300;
  letter-spacing: 0.2rem;
`;

const StyledButton = styled.button`
  background: transparent;
  border: 0.1rem solid #171212;
  padding: 0.5rem;
  cursor: pointer;

  transition: all 0.1s ease-in;

  &:hover {
    background: #171212;
    color: #ffffff;
  }

  &:hover > svg > g {
    fill: palevioletred;
    stroke: #ffffff;
  }
`;

const StyledButtonSmall = styled(StyledButton)`
  padding: 0.5rem;
`;

const StyledButtonLarge = styled(StyledButton)`
  padding: 1rem;
`;

const StyledColumn = styled.div`
  padding: 0rem 0.5rem;
  white-space: nowrap;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  a {
    color: inherit;
  }

  /* Passing in props! */
  width: ${(props) => props.width};
`;

const StyledContainer = styled.div`
  /* height: 100vw; */
  padding: 2rem;

  background: #83a4d4; /* fallback for old browsers */
  background: linear-gradient(to left, #b6fbff, #83a4d4);

  color: #171212;
`;

const StyledItem = styled.div`
  display: flex;
  align-items: center;
  padding-bottom: 0.5rem;
`;

const StyledInput = styled.input`
  border: none;
  border-bottom: 0.1rem solid #171212;
  background-color: transparent;
  font-size: 2.4rem;
`;

const StyledLabel = styled.label`
  border-top: 0.1rem solid #171212;
  border-left: 0.1rem solid #171212;
  padding-left: 0.5rem;
  font-size: 2.4rem;
`;

const StyledSearchForm = styled.form`
  padding: 1rem 0rem 2rem 0rem;
  display: flex;
  align-items: baseline;
`;
