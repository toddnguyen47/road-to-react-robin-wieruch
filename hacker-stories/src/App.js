import React from "react";
import axios from "axios";

import './App.css';

// CONSTANTS
const TypeSetEnum = {
  // Do not start enum at 0 because of JavaScript's strange equality comparison
  // when it comes to 0
  REMOVE_STORIES: 1,
  STORIES_FETCH_INIT: 2,
  STORIES_FETCH_SUCCESS: 3,
  STORIES_FETCH_FAILURE: 4,
};

const title = "React";
const welcome = {
  greeting: "Hey",
  title: title,
};

/**
 *
 * @param {String} title
 */
function getTitle(title) {
  return "| " + title + " |";
}

// const List = (props) => {
function List({ list, onRemoveItem }) {
  return (
    <>
      <div className="item">
        <span style={{ width: '40%' }}>URL</span>
        <span style={{ width: '30%' }}>Author</span>
        <span style={{ width: '10%' }}># Comments</span>
        <span style={{ width: '10%' }}>Points</span>
        <span style={{ width: '10%' }}></span>
      </div>

      <div>
        {list.map((item) => {
          return (
            <Item key={item.objectID} item={item} onRemoveItem={onRemoveItem} />
          );
        })}
      </div>
    </>
  );
}

const Item = ({ item, onRemoveItem }) => {
  return (
    <div className="item">
      <span style={{ width: '40%' }}>
        <a href={item.url}>{item.title}</a>
      </span>
      <span style={{ width: '30%' }}>{item.author}</span>
      <span style={{ width: '10%' }}>{item.num_comments}</span>
      <span style={{ width: '10%' }}>{item.points}</span>
      <span style={{ width: '10%' }}>
        <button type="button" className="button button__small"
          onClick={() => onRemoveItem(item)}>
          Dismiss
        </button>
      </span>
    </div>
  );
};

// /** Passing in a full props javascript object */
// const Search = (props) => {

/** Destructuring in the function parameter! */
const InputWithLabel = ({
  id,
  type = "text",
  value,
  onInputChanged,
  outputString,
  isFocused,
  children,
}) => {
  // (2A) create an inputRef
  const inputRef = React.useRef();

  // (2C) Go into React's lifecycle. Perform the focus when the element renders or
  // when its dependencies change
  React.useEffect(() => {
    // Check if the `.current` property exists
    if (isFocused && inputRef.current) {
      // (2D) Since `inputRef` is passed into the `input` element as the `ref` attribute,
      // `.current` property gives us access to that `input` element
      inputRef.current.focus();
      console.log("Focusing!");
    }
  }, [isFocused]);

  // (1B) Use the callback function
  return (
    <>
      <label htmlFor={id}>{children}</label>
      <input
        /** (2B) The `inputRef` variable is passed to JSX's reserved `ref` attribute.
         * The element instance is assigned to the mutable `current` property
         */
        ref={inputRef}
        id={id}
        type={type}
        onChange={onInputChanged}
        value={value}
      />

      <p>
        {outputString} <strong>'{value}'</strong>
      </p>
    </>
  );
};

/**
 * Store a vallue in localStorage.
 * @param {String} key
 * @param {*} initialState
 * @returns [String, Function]
 */
const useSemiPersistentState = (key, initialState) => {
  const [value, setValue] = React.useState(
    localStorage.getItem(key) || initialState
  );

  /** We will need to monitor both `key` and `value`
   * In case `key` changes outside this hook
   */
  React.useEffect(() => {
    // Store in localStorage every time searchTerm is changed
    localStorage.setItem(key, value);
  }, [key, value]);

  return [value, setValue];
};

const storiesReducer = (state, action) => {
  switch (action.type) {
    case TypeSetEnum.REMOVE_STORIES:
      return {
        ...state,
        data: state.data.filter(
          (story) => action.payload.objectID !== story.objectID
        ),
      };
    case TypeSetEnum.STORIES_FETCH_INIT:
      return {
        ...state,
        isLoading: true,
        isError: false,
      };
    case TypeSetEnum.STORIES_FETCH_SUCCESS:
      return {
        ...state,
        isLoading: false,
        isError: false,
        data: action.payload,
      };
    case TypeSetEnum.STORIES_FETCH_FAILURE:
      return {
        ...state,
        isLoading: false,
        isError: true,
      };
    default:
      throw new Error();
  }
};

const SearchForm = ({ searchTerm, onSearchInput, onSearchSubmit }) => {
  return (
    <>
      <form onSubmit={onSearchSubmit}>
        <InputWithLabel
          id="search"
          value={searchTerm}
          onInputChanged={onSearchInput}
          isFocused={true}
          outputString="The Search Term is: "
        >
          <strong>Search: &nbsp;</strong>
        </InputWithLabel>

        <button type="submit" disabled={!searchTerm}>
          Submit
        </button>
      </form>
    </>
  );
};

/**
 * This is the `App` component.
 * Everything outside is the global space!
 */
const App = () => {
  const [searchTerm, setSearchTerm] = useSemiPersistentState("searchKey", "");

  // const [stories, setStories] = React.useState([]);
  const [stories, dispatchStories] = React.useReducer(storiesReducer, {
    data: [],
    isLoading: false,
    isError: false,
  });

  // (3A) fetch popular tech stories
  const API_ENDPOINT = "https://hn.algolia.com/api/v1/search?query=";

  const handleSearchInput = (event) => {
    setSearchTerm(event.target.value);
  };

  /** Make a new `url` state and a function to update it */
  const [url, setUrl] = React.useState(`${API_ENDPOINT}${searchTerm}`);

  const handleSearchSubmit = (event) => {
    setUrl(`${API_ENDPOINT}${searchTerm}`);
    event.preventDefault();
  };

  const handleRemoveStory = (item) => {
    dispatchStories({
      type: TypeSetEnum.REMOVE_STORIES,
      payload: item,
    });
  };

  const handleFetchStories = React.useCallback(async () => {
    // This check isn't needed technically as the button is disabled,
    // but it does not hurt to check
    if (!url) return;

    dispatchStories({ type: TypeSetEnum.STORIES_FETCH_INIT });
    try {
      const result = await axios.get(url);
      console.log("Fetching!");
      dispatchStories({
        type: TypeSetEnum.STORIES_FETCH_SUCCESS,
        payload: result.data.hits,
      });
    } catch {
      dispatchStories({ type: TypeSetEnum.STORIES_FETCH_FAILURE });
    }
  }, [url]); // (4E) when `searchTerm` changes

  React.useEffect(() => {
    handleFetchStories(); // (4C) invoke the callback using the useEffect() hook
  }, [handleFetchStories]); // (4D) dependency array: depends on any re-defined function handleFetchStories

  return (
    <div className="container">
      <h1 className="headline-primary">
        {welcome.greeting} {welcome.title}
      </h1>

      <h1 className="headline-primary">Hello {getTitle("React")}</h1>

      <SearchForm
        searchTerm={searchTerm}
        onSearchInput={handleSearchInput}
        onSearchSubmit={handleSearchSubmit}
      />

      {stories.isError && <p>Cannot retrieve stories data.</p>}

      {stories.isLoading ? (
        <p>Loading...</p>
      ) : (
          <List list={stories.data} onRemoveItem={handleRemoveStory} />
        )}
    </div>
  );
};

export default App;
