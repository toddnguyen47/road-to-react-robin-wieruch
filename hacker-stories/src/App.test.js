import React from "react";
// import { render } from "@testing-library/react";
import renderer from "react-test-renderer";
// import App, { Item } from "./StyledComponents/App.tsx";
import App, { Item } from "./App.tsx";

// test("renders learn react link", () => {
//   const { getByText } = render(<App />);
//   const linkElement = getByText(/learn react/i);
//   expect(linkElement).toBeInTheDocument();
// });

describe("Item", () => {
  const mockItem = {
    title: "React",
    url: "https://reactjs.org/",
    author: "Jordan Walke",
    num_comments: 3,
    points: 4,
    objectID: 0,
  };

  it("renders all properties", () => {
    let component;
    renderer.act(() => {
      component = renderer.create(<Item item={mockItem} />);
    });

    // expect(component.root.findByType("a").props.href).toEqual(
    //   "https://reactjs.org/"
    // );
    expect(
      component.root.findAllByProps({ href: "https://reactjs.org/" }).length
    ).toEqual(1);

    // expect(component.root.findAllByType("div")[2].props.children).toEqual(
    //   "Jordan Walke"
    // );
    expect(
      component.root.findAllByProps({ children: "Jordan Walke" }).length
    ).toEqual(1);
  });
});
