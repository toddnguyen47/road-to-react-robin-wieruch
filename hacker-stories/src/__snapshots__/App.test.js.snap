// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item renders snapshot 1`] = `
<div
  className="item"
>
  <span
    style={
      Object {
        "width": "40%",
      }
    }
  >
    <a
      href="https://reactjs.org/"
    >
      React
    </a>
  </span>
  <span
    style={
      Object {
        "width": "30%",
      }
    }
  >
    Jordan Walke
  </span>
  <span
    style={
      Object {
        "width": "10%",
      }
    }
  >
    3
  </span>
  <span
    style={
      Object {
        "width": "10%",
      }
    }
  >
    4
  </span>
  <span
    style={
      Object {
        "width": "10%",
      }
    }
  >
    <button
      className="button button__small"
      onClick={[Function]}
      type="button"
    >
      Dismiss
    </button>
  </span>
</div>
`;
